File: lib/to_mdx/parsers/open_document/style/fill_image.rb
# frozen_string_literal: true

require_relative "../base"

module ToMdx
  module Parsers
    module Style
      class FillImage < Base
        class << self
          def fill_images
            @fill_images ||= {}
          end

          def fill_image(name)
            fill_images[name]
          end
        end

        def initialize(element = nil, context = {})
          super(element, context)
          @href = element&.attributes&.[]("xlink:href")
          @name = element&.attributes&.[]("draw:name")
          self.class.fill_images[@name] = self if @name
        end

        def empty?
          @href.empty?
        end

        def to_css
          return {} if empty?

          basename = @context[:basename]
          href = "/img/presentations/#{basename}/#{@href.sub(%r{^Pictures/}, "")}"

          {
            "background-image" => "url('#{href}')",
            "background-repeat" => "no-repeat",
            "background-size" => "cover"
          }
        end

        register_for "draw:fill-image"
      end
    end
  end
end

File: src/parsers/open-document/style/fill_image.ts
import type { ParseContext } from "../../base_element.ts";
import { BaseElement } from "../../base_element.ts";

export class FillImage extends BaseElement {
  private static fillImages = new Map<string, FillImage>();

  static getFillImage(name: string): FillImage | undefined {
    return this.fillImages.get(name);
  }

  private href: string | null;
  private name: string | null;

  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    this.href = this.attr("xlink:href");
    this.name = this.attr("draw:name");
    if (this.name) {
      FillImage.fillImages.set(this.name, this);
    }
  }

  toMdx(): string {
    return "";
  }

  toString(): string {
    return "";
  }

  empty(): boolean {
    return !this.href;
  }

  toCss(): Record<string, string> {
    const basename = this.context("basename");
    if (!basename || !this.href) {
      return {};
    }

    const href = `/img/presentations/${basename}/${this.href.replace(/^Pictures\//, "")}`;
    return {
      "background-image": `url('${href}')`,
      "background-repeat": "no-repeat",
      "background-size": "cover",
    };
  }

  static {
    BaseElement.registerFor.call(this, "draw:fill-image");
  }
}
